watchtower/wtclient/session_queue.go

```diff
@@ -0,0 +1,512 @@
+	})
+}
+
+// ForceQuit idempotently aborts any clean shutdown in progress and returns to
+// he caller after all lingering goroutines have spun down.
+func (q *sessionQueue) ForceQuit() {
+	q.forced.Do(func() {
+		log.Infof("Force quitting session queue %s", q.ID())
+
+		close(q.forceQuit)
+		q.signalUntilShutdown()
+
+		log.Infof("Session queue %s unclean shutdown complete", q.ID())
+	})
+}
+
+// ID returns the wtdb.SessionID for the queue, which can be used to uniquely
+// identify this a particular queue.
+func (q *sessionQueue) ID() *wtdb.SessionID {
+	return &q.cfg.ClientSession.ID
+}
+
+// AcceptTask attempts to queue a backupTask for delivery to the sessionQueue's
+// tower. The session will only be accepted if the queue is not already
+// exhausted and the task is successfully bound to the ClientSession.
+func (q *sessionQueue) AcceptTask(task *backupTask) (reserveStatus, bool) {
+	q.queueCond.L.Lock()
+
+	// Examine the current reserve status of the session queue.
+	curStatus := q.reserveStatus()
+	switch curStatus {
+
+	// The session queue is exhausted, and cannot accept the task because it
+	// is full. Reject the task such that it can be tried against a
+	// different session.
+	case reserveExhausted:
+		q.queueCond.L.Unlock()
+		return curStatus, false
+
+	// The session queue is not exhausted. Compute the sweep and reward
+	// outputs as a function of the session parameters. If the outputs are
+	// dusty or uneconomical to backup, the task is rejected and will not be
+	// tried again.
+	//
+	// TODO(conner): queue backups and retry with different session params.
+	case reserveAvailable:
+		err := task.bindSession(q.cfg.ClientSession)
+		if err != nil {
+			q.queueCond.L.Unlock()
+			log.Debugf("SessionQueue %s rejected backup chanid=%s "+
+				"commit-height=%d: %v", q.ID(), task.chanID,
+				task.commitHeight, err)
+			return curStatus, false
+		}
+	}
+
+	// The sweep and reward outputs satisfy the session's policy, queue the
+	// task for final signing and delivery.
+	q.queue.PushBack(task)
+
+	// Finally, compute the session's *new* reserve status. This will be
+	// used by the client to determine if it can continue using this session
+	// queue, or if it should negotiate a new one.
+	newStatus := q.reserveStatus()
+	q.queueCond.L.Unlock()
+
+	q.queueCond.Signal()
+
+	return newStatus, true
+}
+
+// sessionManager is the primary event loop for the sessionQueue, and is
+// responsible for encrypting and sending accepted tasks to the tower.
+func (q *sessionQueue) sessionManager() {
+	defer close(q.shutdown)
+
+	for {
+		q.queueCond.L.Lock()
+		for q.queue.Front() == nil {
+			q.queueCond.Wait()
+
+			select {
+			case <-q.quit:
+				if q.queue.Len() == 0 {
+					return
+				}
+			case <-q.forceQuit:
+				return
+			default:
+			}
+		}
+		q.queueCond.L.Unlock()
+
+		// Initiate a new connection to the watchtower and attempt to
+		// drain all pending tasks.
+		q.drainBackups()
```

Comment:
since the `Dial` may block indefinitely, we first try to connect, then reacquire the mutex to pull items off. otherwise the client won't be able to continue queueing items while the dial is happening